packages/basic-crawler/src/internals/basic-crawler.ts

```diff
@@ -1130,29 +1147,36 @@ export class BasicCrawler<Context extends CrawlingContext = BasicCrawlingContext
     }
 
     /**
-     * Fetches request from either RequestList or RequestQueue. If request comes from a RequestList
-     * and RequestQueue is present then enqueues it to the queue first.
+     * Initializes both the RequestList and the RequestQueue and then transfers requests from the RequestList
+     * to the RequestQueue using a TandemRequestProvider. This ensures that each request will be processed only once.
      */
-    protected async _fetchNextRequest() {
-        if (!this.requestList) return this.requestQueue!.fetchNextRequest();
-        const request = await this.requestList.fetchNextRequest();
-        if (!this.requestQueue) return request;
-        if (!request) return this.requestQueue.fetchNextRequest();
-
-        try {
-            await this.requestQueue.addRequest(request, { forefront: true });
-        } catch (err) {
-            // If requestQueue.addRequest() fails here then we must reclaim it back to
-            // the RequestList because probably it's not yet in the queue!
-            this.log.error(
-                'Adding of request from the RequestList to the RequestQueue failed, reclaiming request back to the list.',
-                { request },
-            );
-            await this.requestList.reclaimRequest(request);
-            return null;
+    protected async _initializeRequestProviders() {
+        if (this.requestList && this.requestQueue) {
+            // Create a TandemRequestProvider if both RequestList and RequestQueue are provided
+            const requestListAdapter = new RequestListAdapter(this.requestList);
+            const tandemProvider = new TandemRequestProvider(requestListAdapter, this.requestQueue);
+            // Start the background transfer of requests
+            await tandemProvider.startBackgroundTransfer();
```

Comment:
The `startBackgroundTransfer` method shouldn't be part of the public interface, since:
- it exposes internal implementation to the user
- it introduces [temporal coupling](https://en.wikipedia.org/wiki/Sequential_coupling). 
 
I'm not too sure about the background transfer in general (see my comment at `transferAllListRequestsToQueue`), maybe there's a better design choice we can take to solve both of those problems at once.